### DISCUSSION TITLE
replaceall regex does not produce result in trino CLI

### DISCUSSION THREAD

**Discussion by natarajank11** [trinodb/trino#17474] [Q&A] [answered]

**Comment by natarajank11** [comment]
I am working on Trino UDF (version 407). Basically i want to remove '/r/n','/r' and '/n' from the string.

The below code does not work if regex with replaceAll is used in Trino. 

`
import io.airlift.slice.Slice;
import io.airlift.slice.Slices;
import io.trino.spi.function.Description;
import io.trino.spi.function.ScalarFunction;
import io.trino.spi.function.SqlType;
import io.trino.spi.type.StandardTypes;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class RpNewLine
{
    private RpNewLine()
    {}

    @Description("Replace newline code with single-byte space")
    @ScalarFunction("rnc")
    @SqlType(StandardTypes.VARCHAR)
    public static Slice rmcc(@SqlType(StandardTypes.VARCHAR) Slice string)
    {
        String text = string.toStringUtf8();
        String regex = "(\\x0D\\x0A|\\x0D|\\x0A)";
        Pattern p = Pattern.compile(regex);
        Matcher m = p.matcher(text);
        String result = m.replaceAll(" ");

        return Slices.utf8Slice(result);
    }
}

`

If i test in trino, the result is same as the input as shown below.
`trino> select rnc('\nhello\r\nhello\rhello\nhello'); => \nhello\r\nhello\rhello\nhello`



The same logic works fine and produce result with the replace function.
String result = text.replace("\\r\\n", " ").replace("\\r", " ").replace("\\n", " ");
output: hello1 hello2 hello3 hello4

But i don't want to use replace function as i have other regex pattern functionality to solve.
I want to use regex pattern and achieve in replaceall function. please help if i am missing anything.

**Reply by chenjian2664** [reply]
Could you replace the regex to `"(\n\r|\r\n|\r|\n)"` and have a try

**Reply by natarajank11** [reply]
Thank you for your suggestion. Still, there is no expected result,
<img width="1233" alt="image" src="https://github.com/trinodb/trino/assets/24958978/468b77d3-0ef6-44a0-ba5b-284e444e606b">
Also, tried `String regex = "(\n\r|\r\n|\r|\n)";`

For both trial, the results are same.
`trino> select rnc('\nhello\r\nhello\rhello\nhello');
=> \nhello\r\nhello\rhello\nhello`

**Reply by hashhar** [reply]
It's because the `\n` is treated literally in the string literal. String literals in SQL do no have escape characters. You can use a unicode literal (https://trino.io/docs/current/language/types.html#varchar) like `U&'hello\000Aworld'`.

Here `\000A` is the unicode for LF (`\n`).

**Reply by hashhar** [reply]
BTW this isn't something specific to Trino. This is how all databases work - see Postgres https://stackoverflow.com/questions/36028908/postgresql-newline-character and SQL Server https://stackoverflow.com/questions/31057/how-to-insert-a-line-break-in-a-sql-server-varchar-nvarchar-string

**Reply by natarajank11** [reply]
I understand that this could be done in query. However, i would like to form regex pattern in custom UDF to achieve the same behaviour.
I tried `String regex = "(\000A|\000D)";`, still the result is same.

**Reply by hashhar** [reply]
I don't mean for you to change the regex. I mean that in the query the values are treated as literals - that's how SQL works.

So with your existing regex like `String regex = "\\R"` (\R matches all newlines since JDK 8) if you call the function with input like `SELECT rep_newline_code(U&'hello\000Aworld')` you'll get output as `'hello world'`.

**Reply by hashhar** [reply]
If your input has "escape" characters instead of real newlines then your regex would need to be `String regex = "\\r\\n|\\n\\r|\\r|\\n"`.

Also why don't you log the value of `text` to see what input your function actually receives.

**Reply by natarajank11** [reply]
The value for `text` is perfect output. when i try to use udf in trino, then the result are wrong.

**Reply by chenjian2664** [reply, marked as answer]
@natarajank11 How about using the regex `String regex = (\\\\r\\\\n|\\\\n\\\\r|\\\\r|\\\\n)`

**Reply by natarajank11** [reply]
using the regex pattern `(\\\\r\\\\n|\\\\n\\\\r|\\\\r|\\\\n)` is working as expected. 
Thank you!!

**Comment by natarajank11** [comment]
I have similar problem with control characters, this time changed based on the above solution `String regex = "[\\\\p{C}]";`.
However this does not produce result.
Trino result:
`select rm_code('hello\bhello\thello');
=> hellobhellothello
`

Before the change, using `String regex = "[\\p{C}]";`, which produced,
`
select rm_code('hello\bhello\thello');
=> hello\bhello\thello
`

**Reply by natarajank11** [reply]
Just in case, i have created new ticket, https://github.com/trinodb/trino/issues/17614
